rfc_samplers.md

```diff
@@ -0,0 +1,125 @@
+
+Feature space
+-------------
+
+At a high level, a sampler returns a sequence of "clips" where a clip is a batch
+of (non-necessarily contiguous) frames. Based on an analysis of the sampling
+options of other libraries (torchvision, torchmultimodal, and various internal
+libraries), we want to support the following use-cases:
+
+
+**Clip location** (and number thereof)
+
+- Get `num_clips_per_video` random clips: typically called "Random" sampler
+- Get `num_clips_per_video` equally-spaced clips: "Uniform" sampler. Space is
+  index-based.
+- Get `num_clips_per_second` clips per seconds [within the [`start_seconds`,
+  `stop_seconds`] sampling range]: "Periodic" sampler.
+
+**Clip content**
+
+- Clip must be `num_frames_per_clip` frames spaced by `dilation` frames.
+
+Notes:
+
+- There are 3 sampling strategies defining the clip locations: Random, Uniform
+  and Periodic. The names are prone to confusion. "Random" is in fact uniformly
+  random while "Uniform" isn't random. "Uniform" is periodic and "Periodic" is
+  uniformly spaced too.
+- The parametrization of the *Clip Content* dimension is limited and
+  reduces to just `num_frames_per_clip` and the index-based `dilation`.
+- The *Clip Content* dimension is never random. `num_frames_per_clip` is always
+  passed by the user. This makes sense: training typically requires fixed-sized
+  inputs so `num_frames_per_clip` is expected to be constant. The content of a
+  clip (space between frames) is always regular, never random.
+
+
+Proposal
+--------
+
+```py
+from torchcodec.decoders import SimpleVideoDecoder
+from torchcodec import samplers
+
+decoder = SimpleVideoDecoder("./cool_vid")
+
+
+def samplers.clips_at_random_timestamps(
+    decoder,
+    *,
+    num_clips,
+    num_frames_per_clip=1,
+    seconds_between_frames=None,
+) -> FrameBatch
+
+def samplers.clips_at_regular_indices(
```

Comment:
For this and clips_at_regular_timestamps(), both clips and frames within the cips are at regular indices/timestamps respectively, right? Is there a name that would convey to the user that information?

Reply:
The name already convey the fact that clips are at regular indices/timestamps.
For frames within a clip, I don't believe this is necessary to convey. This is not a key aspect of these samplers, and it's only captured in a *single*, optional parameter. We don't convey that in the random sampler either.